compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1279,6 +1280,14 @@ impl<'a> Parser<'a> {
             self.parse_delim_comma_seq(Delimiter::Brace, |p| p.parse_enum_variant()).map_err(
                 |mut e| {
                     e.span_label(id.span, "while parsing this enum");
+                    if self.token == token::Colon {
```

Comment:
I wonder if we can make this heuristic a bit more accurate here if we make the code a bit more complicated (just to avoid spurious errors).

Maybe we can take a snapshot here and try to parse a ty? Or even try to call into whatever code is used to parse struct bodies?

Reply:
Yeah. I believe that we could proactively peek when we encounter the token `enum` past the next ident, past an open `{`, past yet another ident and see if we see a `:`. If so, snapshot and try to parse as a struct. If successful, targeted error and early return, otherwise keep previous error code path.

Reply:
sure, updated the code.